plugin/pkg/proxy/errors.go

```diff
@@ -24,3 +32,19 @@ type Options struct {
 	// HCDomain sets domain for Proxy healthcheck requests
 	HCDomain string
 }
+
+// RcodeToError converts a DNS response code to an error.
+func RcodeToError(rc int) error {
+	switch rc {
+	case 1:
+		return ErrFormatError
+	case 2:
+		return ErrServerFailure
+	case 4:
+		return ErrNotImplemented
+	case 5:
+		return ErrRefused
+	default:
```

Comment:
[performance] Before applying the default case, application will have to check conditions in all other cases.
To make the code faster, consider pre-defining the array/slice  where rcode would be the index and error would be the value. You can also use map, but array/slice would be faster